les/utils/expiredvalue.go

```diff
@@ -0,0 +1,202 @@
+// describes the multiplier applicable for additions and the divider for readouts.
+// If logOffset changes slowly then it saves some expensive operations to not calculate
+// them for each addition and readout but cache this intermediate form for some time.
+// It is also useful for structures where multiple values are expired with the same
+// Expirer.
+type ExpirationFactor struct {
+	Exp    uint64
+	Factor float64
+}
+
+// ExpFactor calculates ExpirationFactor based on logOffset
+func ExpFactor(logOffset Fixed64) ExpirationFactor {
+	return ExpirationFactor{Exp: logOffset.ToUint64(), Factor: logOffset.Fraction().Pow2()}
+}
+
+// Value calculates the expired value based on a floating point base and integer
+// power-of-2 exponent. This function should be used by multi-value expired structures.
+func (e ExpirationFactor) Value(base float64, exp uint64) float64 {
+	res := base / e.Factor
+	if exp > e.Exp {
+		res *= float64(uint64(1) << (exp - e.Exp))
+	}
+	if exp < e.Exp {
+		res /= float64(uint64(1) << (e.Exp - exp))
+	}
+	return res
+}
+
+// value calculates the value at the given moment.
+func (e ExpiredValue) Value(logOffset Fixed64) uint64 {
+	offset := Uint64ToFixed64(e.Exp) - logOffset
+	return uint64(float64(e.Base) * offset.Pow2())
+}
+
+// add adds a signed value at the given moment
+func (e *ExpiredValue) Add(amount int64, logOffset Fixed64) int64 {
+	integer, frac := logOffset.ToUint64(), logOffset.Fraction()
+	factor := frac.Pow2()
+	base := factor * float64(amount)
+	if integer < e.Exp {
+		base /= math.Pow(2, float64(e.Exp-integer))
+	}
+	if integer > e.Exp {
+		e.Base >>= (integer - e.Exp)
+		e.Exp = integer
+	}
+	if base >= 0 || uint64(-base) <= e.Base {
+		e.Base += uint64(base)
+		return amount
+	}
+	net := int64(-float64(e.Base) / factor)
+	e.Base = 0
+	return net
+}
+
+// addExp adds another ExpiredValue
+func (e *ExpiredValue) AddExp(a ExpiredValue) {
+	if e.Exp > a.Exp {
+		a.Base >>= (e.Exp - a.Exp)
+	}
+	if e.Exp < a.Exp {
+		e.Base >>= (a.Exp - e.Exp)
+		e.Exp = a.Exp
+	}
+	e.Base += a.Base
+}
+
+// subExp subtracts another ExpiredValue
+func (e *ExpiredValue) SubExp(a ExpiredValue) {
+	if e.Exp > a.Exp {
+		a.Base >>= (e.Exp - a.Exp)
+	}
+	if e.Exp < a.Exp {
+		e.Base >>= (a.Exp - e.Exp)
+		e.Exp = a.Exp
+	}
+	if e.Base > a.Base {
+		e.Base -= a.Base
+	} else {
+		e.Base = 0
+	}
+}
+
+// Expirer changes logOffset with a linear rate which can be changed during operation.
+// It is not thread safe, if access by multiple goroutines is needed then it should be
+// encapsulated into a locked structure.
+// Note that if neither SetRate nor SetLogOffset are used during operation then LogOffset
+// is thread safe.
+type Expirer struct {
+	logOffset  Fixed64
+	rate       float64
+	lastUpdate mclock.AbsTime
+}
+
+// SetRate changes the expiration rate which is the inverse of the time constant in
+// nanoseconds.
+func (e *Expirer) SetRate(now mclock.AbsTime, rate float64) {
```

Comment:
If it's called first(initialization), we shouldn't accumulate offset right?

Reply:
We don't because e.rate is 0.

Reply:
I mean the L182 in `les/lespay/client/valueTracker.go`, we first time initialize the expired but current system time and a default rate.  Then the offset is accumulated from time(0)

Reply:
Btw I test it manually, for the first-time initialization, the time diff is `diff 205996268056394` and the rate is ` rate 2.777777777777778e-16`.

The time diff should be zero, and the rate is definitely too low.

Reply:
The rate seems to be fine to me, it should be the inverse of the time constant in nanoseconds (which is huge).
The initial `dt` at the first `SetRate` is indeed a meaningless value because `e.lastUpdate` is not initialized but it is multiplied by the previous rate (`e.rate`) which is zero so `dt` does not matter at all here. Next time when `e.rate` is non-zero the `dt` value will also be meaningful.

Reply:
What's the `previous rate`? From the code there is only one rate specifed `1/float64(time.Hour*1000)`

Reply:
Oh, sorry, it's `e.rate`. NVM